bindings.go

```diff
@@ -388,3 +390,141 @@ func (n Node) PrevNamedSibling() *Node {
 func (n Node) Edit(i EditInput) {
 	C.ts_node_edit(&n.c, i.c())
 }
+
+// QueryErrorType - value that indicates the type of QueryError.
+type QueryErrorType int
+const (
+	QueryErrorNone QueryErrorType = iota
+	QueryErrorSyntax
+	QueryErrorNodeType
+	QueryErrorField
+	QueryErrorCapture
+)
+
+// QueryError - if there is an error in the query,
+// then the Offset argument will be set to the byte offset of the error,
+// and the Type argument will be set to a value that indicates the type of error.
+type QueryError struct {
+	Offset uint32
+	Type QueryErrorType
+}
+func (qe *QueryError) Error() string {
+	switch qe.Type {
+	case QueryErrorNone:
+		return ""
+
+	case QueryErrorSyntax:
+		return fmt.Sprintf("syntax error (offset: %d)", qe.Offset)
+
+	case QueryErrorNodeType:
+		return fmt.Sprintf("node type error (offset: %d)", qe.Offset)
+
+	case QueryErrorField:
+		return fmt.Sprintf("field error (offset: %d)", qe.Offset)
+
+	case QueryErrorCapture:
+		return fmt.Sprintf("capture error (offset: %d)", qe.Offset)
+
+	default:
+		return fmt.Sprintf("unknown error (offset: %d)", qe.Offset)
+	}
+}
+
+// Query API
+type Query struct { c *C.TSQuery }
+
+// NewQuery creates a query by specifying a string containing one or more patterns.
+func NewQuery(pattern []byte, lang *Language) (*Query, *QueryError) {
+	var (
+		erroff C.uint32_t
+		errtype C.TSQueryError
+	)
+
+	c := C.ts_query_new(
+		(*C.struct_TSLanguage)(lang.ptr),
+		(*C.char)(C.CBytes(pattern)),
+		C.uint32_t(len(pattern)),
+		&erroff,
+		&errtype,
+	)
+	if errtype != C.TSQueryError(QueryErrorNone) {
+		return nil, &QueryError{Offset: uint32(erroff), Type: QueryErrorType(errtype)}
+	}
+
+	q := &Query{c}
+	runtime.SetFinalizer(q, deleteQuery)
+
+	return q, nil
+}
+func deleteQuery(q *Query) {
+	C.ts_query_delete(q.c)
+}
+
+// QueryCursor carries the state needed for processing the queries.
+type QueryCursor struct {c *C.TSQueryCursor }
+
+// NewQueryCursor creates a query cursor.
+func NewQueryCursor() *QueryCursor {
+	qc := &QueryCursor{C.ts_query_cursor_new()}
+	runtime.SetFinalizer(qc, deleteQueryCursor)
+
+	return qc
+}
+func deleteQueryCursor(qc *QueryCursor) {
+	C.ts_query_cursor_delete(qc.c)
+}
+
+// Exec executes the query on a given syntax node.
+func (qc *QueryCursor) Exec(q *Query, n *Node) {
+	C.ts_query_cursor_exec(qc.c, q.c, n.c)
+}
+
+// QueryCapture
+type QueryCapture struct {
+	Index uint32
+	Node *Node
+}
+
+// QueryMatch - you can then iterate over the matches.
+type QueryMatch struct {
+	Id uint32
+	PatternIndex uint16
+	Captures []QueryCapture
+}
+
+// NextMatch iterates over matches.
+// This function will return (nil, false) when there are no more matches.
+// Otherwise, it will populate the QueryMatch with data
+// about which pattern matched and which nodes were captured.
+func (qc *QueryCursor) NextMatch() (*QueryMatch, bool) {
```

Comment:
Do we need a bool return here? A `!= nil` check could be good enough.

Reply:
It's to keep original C-interface but also look at native go:
maps lookup has a boolean information of key exists,
channel contains extra information if it was closed.
I think `bool` here is more handy for iterator and it's very go-like.